### PR TITLE
Multipage document creation

### PR DESCRIPTION
**Pull request by kikettas**
For multi-page analysis, several methods have been added into the `GINIDocumentTaskManager` class, along with some document related methods that were in the `GINIDocument` class.
FYI, methods which have the `cancellationToken` parameter don't come from last public release but from a previous PR.

### How to test
Not possible yet since it is only available for local development.

### Merging
Automatic.

Issue #69

### CODE REVIEW

File: Gini-iOS-SDK/GINIDocument.h [hunk 1]
```diff
@@ -53,21 +56,27 @@ typedef NS_ENUM(NSUInteger, GiniDocumentSourceClassification) {
...
 @property GiniDocumentSourceClassification sourceClassification;
+/// Links to related resources, such as extractions, document, processed or layout.
+@property (readonly) GINIDocumentLinks *links;
+/// (Optional) Array containing the path of every parent
+@property (readonly) NSArray<NSString *> *parents;
+/// (Optional) Array containing the path of every partial document
+@property (readonly) NSArray<NSString *> *partialDocuments;
 /// A `BFTask*` resolving to a mapping with extractions (extraction name as key).
-@property (readonly) BFTask *extractions;
+@property (readonly) BFTask *extractions __attribute__((unavailable("use `GINIDocumentTaskManager.getExtractionsForDocument:` method instead")));
```
Comment (nsobadjiev) [inline Gini-iOS-SDK/GINIDocument.h]: Don't you have to deprecate it first?

File: Gini-iOS-SDK/GINIDocumentTaskManager.h [hunk 2]
```diff
@@ -132,46 +101,55 @@
...
+                                     fromData:(NSData *)data
+                                      docType:(NSString *)docType
+                            cancellationToken:(BFCancellationToken *)cancellationToken;
 
 /**
- * Saves updates on the extractions.
+ * Creates a new composite document
  *
- * Updating extractions is called "Submitting feedback" in the Gini API documentation.
+ * See the [Gini API documentation](Add documentation link).
```
Comment (nsobadjiev) [inline Gini-iOS-SDK/GINIDocumentTaskManager.h]: Link please...

File: Gini-iOS-SDK/GINIDocumentTaskManager.m [hunk 3]
```diff
@@ -17,18 +17,18 @@
     return [originalTask continueWithBlock:^id(BFTask *task) {
         if (task.error && [task.error.domain isEqualToString:NSURLErrorDomain]) {
             switch(task.error.code) {
-                // HTTP #404
+                    // HTTP #404
```
Comment (nsobadjiev) [inline Gini-iOS-SDK/GINIDocumentTaskManager.m]: Why did you add this tab? Shouldn't it align with the case statement?

File: Gini-iOS-SDK/GINIDocumentTaskManager.m [hunk 4]
```diff
@@ -61,76 +61,127 @@ - (instancetype)initWithAPIManager:(GINIAPIManager *)apiManager {
...
-}
+- (BFTask *)createDocumentWithFilename:(NSString *)fileName
+                              fromData:(NSData *)data
+                               docType:(NSString *)docType
+                     cancellationToken:(BFCancellationToken *)cancellationToken {
+    NSParameterAssert([fileName isKindOfClass:[NSString class]]);
+    NSParameterAssert([data isKindOfClass:[NSData class]]);
+    
+    BFTask *createTask = [[_apiManager uploadDocumentWithData:data
+                                                  contentType:@"image/jpeg"
```
Comment (nsobadjiev) [inline Gini-iOS-SDK/GINIDocumentTaskManager.m]: Are you sure it's always a jpeg?

File: Gini-iOS-SDK/GINIDocumentTaskManager.m [hunk 5]
```diff
@@ -61,76 +61,127 @@ - (instancetype)initWithAPIManager:(GINIAPIManager *)apiManager {
...
+    NSString* lastContentTypeComponent = [[contentType componentsSeparatedByString:@"/"] lastObject];
+    NSString* concreteType = @"";  // i.e: jpeg
+    if (lastContentTypeComponent != nil && [lastContentTypeComponent length] > 0) {
+        concreteType = lastContentTypeComponent;
+    }
+    contentType = [NSString stringWithFormat:@"application/vnd.gini.v2.partial+%@", concreteType];
+    
     BFTask *createTask = [[_apiManager uploadDocumentWithData:data
-                                                  contentType:@"image/jpeg"
+                                                  contentType:contentType
```
Comment (nsobadjiev) [inline Gini-iOS-SDK/GINIDocumentTaskManager.m]: Here, you pass the document type and use it. But in the previous method it was hardcoded to jpeg.

File: Gini-iOS-SDK/GINIDocumentTaskManager.m [hunk 6]
```diff
@@ -61,76 +61,127 @@ - (instancetype)initWithAPIManager:(GINIAPIManager *)apiManager {
...
                                             cancellationToken:cancellationToken] continueWithSuccessBlock:^id(BFTask *task) {
-        return [GINIDocument documentFromAPIResponse:task.result withDocumentManager:self];
+        return [GINIDocument documentFromAPIResponse:task.result];
     }];
     return GINIhandleHTTPerrors(createTask);
 }
 
-- (BFTask *)updateDocument:(GINIDocument *)document {
-    return [self updateDocument:document cancellationToken:nil];
+- (BFTask *)createCompositeDocumentWithPartialDocumentsInfo:(NSArray<NSDictionary<NSString *,id> *> *)partialDocumentsInfo
```
Comment (nsobadjiev) [inline Gini-iOS-SDK/GINIDocumentTaskManager.m]: `(NSArray<NSDictionary<NSString *,id> *> *)` is a bit difficult to look at. Why isn't it an object?

File: Gini-iOS-SDK/GINIDocumentTaskManager.m [hunk 7]
```diff
@@ -141,7 +192,38 @@ - (BFTask *)deleteDocument:(GINIDocument *)document
...
+            return [self deleteCompositeDocumentWithId:document.documentId cancellationToken:cancellationToken];
+        }];
+    }];
+}
+
+- (BFTask *)deleteDocumentsWithUrls:(NSArray<NSString*> *)urls cancellationToken:(BFCancellationToken *)cancellationToken {
+    NSMutableArray<BFTask *>* deleteTasks = [NSMutableArray new];
+    
+    for (NSString* url in urls) {
+        NSString* documentId = [[url componentsSeparatedByString:@"/"] lastObject];
```
Comment (nsobadjiev) [inline Gini-iOS-SDK/GINIDocumentTaskManager.m]: It would have been nice if you could avoid code like this. It kind of makes assumptions about how URLs in the service are formed. It is a little fragile.

File: Gini-iOS-SDK/GINIDocumentTaskManager.m [hunk 8]
```diff
@@ -185,55 +267,32 @@ - (BFTask *)privatePollDocumentWithId:(NSString *)documentId
...
-
-- (BFTask *)getIncubatorExtractionsForDocument:(GINIDocument *)document
-                             cancellationToken:(BFCancellationToken *)cancellationToken {
-    NSParameterAssert([document isKindOfClass:[GINIDocument class]]);
+    NSMutableDictionary *filteredUpdatedExtractions = [NSMutableDictionary new];
     
-    BFTask *extractionsTask = [self createExtractionsForGetTask:[_apiManager getIncubatorExtractionsForDocument:document.documentId
-                                                                                              cancellationToken:cancellationToken]];
-    return GINIhandleHTTPerrors(extractionsTask);
+    NSArray *keys = @[@"paymentReference", @"iban", @"bic", @"amountToPay", @"paymentRecipient", @"paymentPurpose"];
```
Comment (nsobadjiev) [inline Gini-iOS-SDK/GINIDocumentTaskManager.m]: This is how you work with feedback, right? I know e only get feedback for these specific fields, but for someone without the domain knowledge it would be difficult to understand this code. Maybe you can add a comment and possibly extract these fields somewhere?

File: Gini-iOS-SDK/GINIAPIManager.h [hunk 9]
```diff
@@ -208,6 +208,28 @@ typedef NS_ENUM(NSUInteger, GiniAPIResponseType){
                            docType:(NSString *)docType
                  cancellationToken:(BFCancellationToken *) cancellationToken;
 
+
+/**
+ * Creates a new composite document
+ *
+ * See the [Gini API Documentation](Add documentation link).
```
[the remark this answers is not in the document]
Comment (kikettas) [reply Gini-iOS-SDK/GINIAPIManager.h]: There is no link because the API documentation has not been written yet.

File: Gini-iOS-SDK/GINIAPIManager.h [hunk 10]
```diff
@@ -208,6 +208,28 @@ typedef NS_ENUM(NSUInteger, GiniAPIResponseType){
...
+ *
+ * @param partialDocumentsInfo  Array containing the partial documents info. More info can be found [here](Add here documentation link)
```
[the remark this answers is not in the document]
Comment (kikettas) [reply Gini-iOS-SDK/GINIAPIManager.h]: Answered above

File: Gini-iOS-SDK/GINIAPIManager.m [hunk 11]
```diff
@@ -203,6 +203,32 @@ - (BFTask *)uploadDocumentWithData:(NSData *)documentData
...
+- (BFTask *)createCompositeDocumentWithPartialDocumentsInfo:(NSArray<NSDictionary<NSString *,id> *> *)partialDocumentsInfo
+                                                   fileName:(NSString *)fileName
+                                                    docType:(NSString *)docType
+                                          cancellationToken:(BFCancellationToken *)cancellationToken {   
+    NSDictionary* dict = @{@"partialDocuments": partialDocumentsInfo};
+    NSData *jsonDataFormatted = [self jsonDataFormattedFromDictionary:dict];
+    
+    NSString *urlString = [NSString stringWithFormat:@"documents/?filename=%@", stringByEscapingString(fileName)];
+    if (docType) {
+        urlString = [urlString stringByAppendingString:[NSString stringWithFormat:@"&doctype=%@", stringByEscapingString(docType)]];
```
[the remark this answers is not in the document]
Comment (kikettas) [reply Gini-iOS-SDK/GINIAPIManager.m]: Honestly, I reused what was used in the other methods to be consistent. It is more prone to errors, but easier to read.

File: Gini-iOS-SDK/GINIDocumentTaskManager.h (hunk 2 above)
Comment (kikettas) [reply Gini-iOS-SDK/GINIDocumentTaskManager.h]: Answered above.

File: Gini-iOS-SDK/GINIDocumentTaskManager.m (hunk 3 above)
Comment (kikettas) [reply Gini-iOS-SDK/GINIDocumentTaskManager.m]: I don't know why Xcode did this, but I can change it.

File: Gini-iOS-SDK/GINIDocumentTaskManager.m (hunk 4 above)
Comment (kikettas) [reply Gini-iOS-SDK/GINIDocumentTaskManager.m]: I added this method before creating the mimeType helper. Nice catch!

File: Gini-iOS-SDK/GINIDocumentTaskManager.m (hunk 5 above)
Comment (kikettas) [reply Gini-iOS-SDK/GINIDocumentTaskManager.m]: Answered above.

File: Gini-iOS-SDK/GINIDocumentTaskManager.m (hunk 7 above)
Comment (kikettas) [reply Gini-iOS-SDK/GINIDocumentTaskManager.m]: I thought about this, but at some point you have to get the id from the url (since you always get the url). I didn't come with a better idea, do you have a good one?

File: Gini-iOS-SDK/GINIAPIManager.m [hunk 12]
```diff
@@ -203,6 +203,32 @@ - (BFTask *)uploadDocumentWithData:(NSData *)documentData
     } cancellationToken:cancellationToken];
 }
 
+- (BFTask *)createCompositeDocumentWithPartialDocumentsInfo:(NSArray<NSDictionary<NSString *,id> *> *)partialDocumentsInfo
```
[the remark this answers is not in the document]
Comment (kikettas) [reply Gini-iOS-SDK/GINIAPIManager.m]: Partial documents info is just a json made of different partial document ids and additional parameters like rotationDelta. Maybe _Info_ is not the best word, but with the link to the API it should be clear.

File: Gini-iOS-SDK/GINIDocument.h (hunk 1 above)
Comment (kikettas) [reply Gini-iOS-SDK/GINIDocument.h]: I was thinking to make them `unavailable` since it is a major version, but we can actually deprecate them now and remove them in a future version.

File: Gini-iOS-SDK/GINIDocumentTaskManager.h [hunk 13]
```diff
@@ -132,46 +101,55 @@
...
+ * @param fileName              The filename of the document.
+ * @param docType               (Optional) A doctype hint. This optimizes the processing at the Gini API.
+ *                              for a list of possibles doctypes.
+ * @param cancellationToken     Cancellation token used to cancel the current task.
+ *
+ * @returns                     A`BFTask*` that will resolve to a NSString containing the created document's ID.
  */
-- (BFTask *)updateDocument:(GINIDocument *)document;
+
+- (BFTask *)createCompositeDocumentWithPartialDocumentsInfo:(NSArray<NSDictionary<NSString *, id>*>*)partialDocumentsInfo
```
Comment (a-szotyori) [inline Gini-iOS-SDK/GINIDocumentTaskManager.h]: I would prefer a more stricter public api. This method is extensible which is great. The downside is, that it's harder to use: what would I as a client have to put into the dictionary?

By narrowing it down to what we need right now would make it imho easier to use. We need a list of documents and rotation deltas.

For example:
```
- (BFTask *)createCompositeDocumentWithPartialDocumentsInfo:(NSArray<PartialDocumentInfo*>*)partialDocumentsInfo
```

`PartialDocumentInfo` could contain the `GINIDocument` (or just the document url) and the rotation delta. Using a container object is more verbose and looks nicer to me than using a dictionary inside of an array. This also maintains the document order like in your implementation.

File: Gini-iOS-SDK/GINIAPIManager.m [hunk 14]
```diff
@@ -203,6 +203,32 @@ - (BFTask *)uploadDocumentWithData:(NSData *)documentData
...
+                                                   fileName:(NSString *)fileName
+                                                    docType:(NSString *)docType
+                                          cancellationToken:(BFCancellationToken *)cancellationToken {   
+    NSDictionary* dict = @{@"partialDocuments": partialDocumentsInfo};
```
Comment (a-szotyori) [inline Gini-iOS-SDK/GINIAPIManager.m]: This method is public (correct me if I'm wrong). In which case I find it dangerous to expect clients to pass in a `partialDocumentsInfo` which contains correct key names and values.

In my comment for the method declaration in the header file I suggested passing in an array of `PartialDocumentInfo` objects. Using something in the lines of that to separate the required json structure and the client supplied argument would avoid messing up the json structure. We would need to a little more overhead to serialize to json, but would be in control of it.
